README.md

# QUICK LINKS ![logo][logo] #

## Description

> Tired or typing and re-typing URLs in your Address bar. Get Quick Links

This Mozilla Firefox Addon adds iconic links to the browser addon bar (status bar) that you can click on to *quickly* open the allocated URL in a new tab.


## Homepage

More details can be found in the addon [site][addon-site]


## Details

|Aspect|Detail|
|---------|---------:|
|*Current Version*| 0.2.0|
|*Firefox Compatibility*| 21.0a - 32.0|
|*Platforms*| Linux/Windows/MacOSx|
|*Review Status*|N/A|
|*Languages*|English(US), Swahili, French|
|*Last Updated*|December 27, 2014|


## Installation

### 1. Mozilla.org (recommended)

The Addon is listed on the Mozilla.org site [here][download-mozilla].

### 2. Build with Mozilla addon-sdk-1.17

1. Download ZIP or clone this repo: `git clone https://github.com/GochoMugo/Quick-Links.git`
2. Build and Install the addon: `make`
    - This will automatically install the addon sdk (if `cfx` is __not__ available).
    - Builds the addon and installs it.
    - If it fails, you may require `sudo make` to make cfx be installed properly.
    - The `cfx` command will be available to you from your terminal.

> You may find more information on building addons [here][addon-sdk]


## Contribution

In case of a bug or feature request, you can create an [issue][issues] and let us look into it as soon as possible.

If you wish to translate this addon to another language, please view the [translations][translations] page.

From [here][correspondence] anybody can report a bug, suggest a change or simply say a thumbs up to the Developer. Looking forward to hear from you :-)

All bugs, updates and changes are posted at the [page][bug-report-page]. 


## Credits

A list of all the contributors to this addon.

1. <a href='https://gochomugo.github.io'>GochoMugo</a>
2. <a href='mailto:mybeautifulrat@gmail.com'>Magali de Morangies</a>

Contribute to this project and see your name here...


## License

__The MIT License (MIT)__

Copyright (c) 2014 Gocho Mugo <mugo@forfuture.co.ke>


[logo]:https://gochomugo.github.io/Quick-Links/images/logo.png "Quick Links"
[addon-site]:https://gochomugo.github.io/Quick-Links/index.html "Go to Addon Site"
[download-mozilla]:https://addons.mozilla.org/en-US/firefox/addon/quick-links/ "Download from mozilla.org"
[addon-sdk]:https://developer.mozilla.org/en-US/Add-ons/SDK "Tutorials on the addon sdk"
[issues]:https://github.com/GochoMugo/Quick-Links/issues "Issues on Github"
[translations]:https://gochomugo.github.io/Quick-Links/translations.html "Translate to another Language"
[correspondence]:https://gochomugo.github.io/Quick-Links/bugs_issues.html#correspondence "Report a bug, Suggest a change or Say a Thumbs up"
[bug-report-page]:https://gochomugo.github.io/Quick-Links/bugs_issues.html "View reported bugs and suggested changed"
[license-page]:https://gochomugo.github.io/Quick-Links/license.html "View License"


changelog
Key
----
+ Improvement
- Backwards-incompatible Change
= Bug Fix


[0.3.0]
=====================


[0.2.0]
=====================
1.  +  Clicking icons in the add panel opens a new tab with its associated url loaded.
2.  +  French translation added. <By Magali de Morangies>
3.  =  Notification shown only major and minor version bumps occur. Ignores patch version bumps.


[0.1.0]
====================

1.  +  Overhaul of data structure with an object-oriented feel
        *   Links can be easily added to the addon by:
        * add a link entry into "presets.js"
        * use the id (from the newly-created entry above) in locale files to provide a translation

2.  +  Added a background image on the panel
3.  +  Added a hotkey for quickly creating a quick link
4.  +  More preset icons: (3 new)
        * Foursquare
        * JSFiddle
        * Trello

5.  =  Successful migration from v0.0.2 to v0.1.0


[0.0.2]
====================

1.  +  More light-weight
        * Unrequired Jquery Files have been deleted
        * Non-minified css and js in both Bootstrap and Font Awesome Library have been removed

2.  +  Semantic Versioning Enforcement
        * First-run shows the correct addon version

3.  +  Compatibility specification in First-run.html corrected for accuracy
4.  =  Links in first-run.html removed
        * The images were wrapped in Links with href attr pointing to the images dir
        * However the Link was "dead". Image didn't show up

5.  +  Font-Awesome upgraded to 4.1.0
6.  +  More preset icons:  (15 new)
        * Apple
        * Codepen
        * Drupal
        * Git
        * Google
        * Pinterest
        * SoundCloud
        * Stack Exchange
        * Stack Overflow
        * Steam
        * Tumblr
        * Vimeo
        * Windows
        * Wordpress
        * Yahoo

7.  +  More custom icons: (2 new)
        * comments
        * cutlery
        * flag
        * institution
        * paper-plane


[0.0.1]
====================
1.  +  Icons added
        * icon.png(48 x 48) and icon64.png(64 x 64) have been added

2.  +  Package.json updated for icons and semantic versioning
3.  =  Holder.js in Bootstrap is removed
4.  =  HTML Insertion
        *Security measure
        *Ineffeciency
        *All .innerHTML += operations are now handled with jQuery's .append() a$
        *All .innerHTML -= operations are now handled with jQuery's .remove()
        *No more usanitized html. The elements required are created using docum$

5.  +   Compatibility Tested
            * 21.0a - 32.0


src/lib/main.js
/**
 * QuickLinks -- Addon main module
 *
 * Copyright (c) 2014-2015 GochoMugo <mugo@forfuture.co.ke>
 * See LICENSE in project root directory
 */


// built-in modules
import buttons from "sdk/ui/button/action";
import hotkeys from "sdk/hotkeys";
import panels from "sdk/panel";
import self from "sdk/self";
import storage from "sdk/simple-storage";
import tabs from "sdk/tabs";
import widgets from "sdk/widget";


// own modules
// import migrations from "./migrations";
import presets from "./presets";
import utils from "./utils";


// module variables
let links, activeLinks, panel, widget, hotkey, button; // eslint-disable-line no-unused-vars


// Run migrations
// migrations.run();


function getLinks(l) {
  widget.width = 40;
  links = l;
  activeLinks = [];
  links.forEach(function(link) {
    if (link.active) {
      activeLinks.push(link);
      widget.width += 15;
      if (link.keyword) {
        widget.width += link.keyword.length * 12;
      }
    }
  });
  storage.links = links; // Storing links in system
  return activeLinks;
}


// Panel for Settings
panel = panels.Panel({
  width: 600,
  height: 520,
  contentURL: self.data.url("add.html"), // html document
  contentScriptFile: [
    self.data.url("vendor/jquery/jquery-1.10.2.js"),
    self.data.url("vendor/bootstrap/dist/js/bootstrap.min.js"),
    self.data.url("js/add.js"),
  ],
  contentScriptWhen: "ready",
  onHide: function () {
    panel.postMessage({ aim: "close" });
  },
  onMessage: function (message) {
    /*
     * Receiving messages:
     *   "close": hide the panel
     *   "translate":  translate text using locales
     *   "links": links received from the panel. Tell widget to update addon
     *   "openLink": iconic link on panel has been clicked. Open a tab.
     */
    switch (message.aim) {
    case "close":
      panel.hide();
      break;
    case "translate":
      panel.postMessage(utils.translate(message.content));
      break;
    case "links":
      widget.postMessage({
        aim: "update",
        content: getLinks(message.content),
      });
      break;
    case "linkOpen":
      tabs.open(message.content);
      panel.hide();
      break;
    }
  },
});


// Widget on the Addon bar
widget = widgets.Widget({
  id: "QL_widget",
  label: "Click to open...",
  contentURL: self.data.url("main.html"),
  contentScriptFile: [
    self.data.url("jquery/jquery-1.10.2.js"),
    self.data.url("js/main.js"),
  ],
  contentScriptWhen: "ready",
  onAttach: function () {
    /*
     * If links had been stored, we placed in the .links variable. If no links had
     * been stored we load the presets from the "presets" module and place
     * them in the variable
     * We then update widget and the panel
     */
    if (!storage.links) {
      links = presets.get();
    } else {
      links = storage.links;
    }
    widget.postMessage({
      aim: "update",
      content: getLinks(links),
    });
    panel.postMessage({
      aim: "links",
      content: links,
    });
  },
  onMessage: function (message) {
    /*
     * Receives all messages sent to the Widget.
     *   "open_panel": open the panel
     *   "url": open new tab with url
     *   "presets": receive presets from presets.js
     *   "translate": presets require translation
     */
    switch (message.aim) {
    case "open_panel":
      panel.show();
      break;
    case "url":
      tabs.open(message.content);
      break;
    }
  },
});


/*
 * Registering the hotkey {Ctrl-Alt-Q}
 *
 * This will take note of the current tab"s URL and title and fill them in the panel
 * while showing it
 */
 hotkey = hotkeys.Hotkey({
  combo: "accel-alt-q",
  onPress: function () {
    /*
     * Gets the active tab. Extracts the title and URL
     * Sends message on behalf of panel to have the second tab activated and
     * the appropriate textboxes be filled with the title and url
     * Opens the Panel after that
     */
    var activeTab = tabs.activeTab;
    var title = activeTab.title;
    var url = activeTab.url;
    panel.postMessage({
      aim: "bookmark",
      content: {
        title,
        url,
      },
    });
   panel.show();
  },
});


/*
 * Ensuring OverQuota is handled
 */
storage.on("OverQuota", function(){
  panel.postMessage("OverQuota");
  utils.notify(
    "Memory Exhaustion",
    "You have used up memory reserved for Quick Links. You may delete some icons.",
    function () {
      panel.show();
    }
  );
});



button = buttons.ActionButton({
  id: "mozilla-link",
  label: "Visit Mozilla",
  icon: {
    "16": self.data.url("img/icon.png"),
    "32": self.data.url("img/icon.png"),
    "64": self.data.url("img/icon64.png"),
  },
  onClick: function() {
    panel.show();
  },
});
